# Census bureau database

In [1]:
import numpy as np
import pandas as pd

In [2]:
train_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', header = None)
test_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test' ,skiprows = 1, header = None)

In [3]:
print(train_set.shape, test_set.shape)

(32561, 15) (16281, 15)


In [4]:
train_set.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [5]:
train_set.iloc[0, 14]

' <=50K'

In [6]:
test_set.head(5)

0           1       2              3   4                    5   \
0  25     Private  226802           11th   7        Never-married   
1  38     Private   89814        HS-grad   9   Married-civ-spouse   
2  28   Local-gov  336951     Assoc-acdm  12   Married-civ-spouse   
3  44     Private  160323   Some-college  10   Married-civ-spouse   
4  18           ?  103497   Some-college  10        Never-married   

                   6           7       8        9     10  11  12  \
0   Machine-op-inspct   Own-child   Black     Male     0   0  40   
1     Farming-fishing     Husband   White     Male     0   0  50   
2     Protective-serv     Husband   White     Male     0   0  40   
3   Machine-op-inspct     Husband   Black     Male  7688   0  40   
4                   ?   Own-child   White   Female     0   0  30   

               13       14  
0   United-States   <=50K.  
1   United-States   <=50K.  
2   United-States    >50K.  
3   United-States    >50K.  
4   United-States   <=50K.

In [6]:
test_set.iloc[4, 1]

' ?'

In [7]:
test_set.iloc[0, 14]

' <=50K.'

In [8]:
test_set.iloc[2, 14]

' >50K.'

few problems we can see by now:-<br>
1- don't have column header for our data.<br>
2- ? mark present in our data.<br>
3- categorical target variable.

### 1- putting column in datasets.

In [9]:
col_labels = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation','relationship',
              'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week','native_country', 'wage_class']

* Putting columns header on both datasets.

In [10]:
train_set.columns = col_labels
test_set.columns = col_labels

In [11]:
train_set.head(3)

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   

        marital_status          occupation    relationship    race    sex  \
0        Never-married        Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse     Exec-managerial         Husband   White   Male   
2             Divorced   Handlers-cleaners   Not-in-family   White   Male   

   capital_gain  capital_loss  hours_per_week  native_country wage_class  
0          2174             0              40   United-States      <=50K  
1             0             0              13   United-States      <=50K  
2             0             0              40   United-States      <=50K

### 2- identifying any null value present in datasets.

In [12]:
train_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education_num     32561 non-null int64
marital_status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital_gain      32561 non-null int64
capital_loss      32561 non-null int64
hours_per_week    32561 non-null int64
native_country    32561 non-null object
wage_class        32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [13]:
test_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16281 entries, 0 to 16280
Data columns (total 15 columns):
age               16281 non-null int64
workclass         16281 non-null object
fnlwgt            16281 non-null int64
education         16281 non-null object
education_num     16281 non-null int64
marital_status    16281 non-null object
occupation        16281 non-null object
relationship      16281 non-null object
race              16281 non-null object
sex               16281 non-null object
capital_gain      16281 non-null int64
capital_loss      16281 non-null int64
hours_per_week    16281 non-null int64
native_country    16281 non-null object
wage_class        16281 non-null object
dtypes: int64(6), object(9)
memory usage: 1.9+ MB


it is showing there is no missing value in both the datasets. but we know there is some missing value but they are in the form of '?'. so we first have to replace '?' by null value then later we'll try to impute them.

In [14]:
train_set = train_set.replace(' ?', np.nan)
test_set = test_set.replace(' ?', np.nan)

In [15]:
test_set.head(5)

age   workclass  fnlwgt      education  education_num       marital_status  \
0   25     Private  226802           11th              7        Never-married   
1   38     Private   89814        HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951     Assoc-acdm             12   Married-civ-spouse   
3   44     Private  160323   Some-college             10   Married-civ-spouse   
4   18         NaN  103497   Some-college             10        Never-married   

           occupation relationship    race      sex  capital_gain  \
0   Machine-op-inspct    Own-child   Black     Male             0   
1     Farming-fishing      Husband   White     Male             0   
2     Protective-serv      Husband   White     Male             0   
3   Machine-op-inspct      Husband   Black     Male          7688   
4                 NaN    Own-child   White   Female             0   

   capital_loss  hours_per_week  native_country wage_class  
0             0              40   United-States     <=50K.  
1             0              50   United-States     <=50K.  
2             0              40   United-States      >50K.  
3             0              40   United-States      >50K.  
4             0              30   United-States     <=50K.

if there is a null value in target variable i.e. 'wage_class' then there is no use of that perticular row so we should dropout that row.

In [16]:
train_set.iloc[:, 14].isnull().any()

False

In [17]:
test_set.iloc[:, 14].isnull().any()

False

### 3- Seperating the target variable from rest of the dataset.

In [18]:
X_train = train_set.drop(['wage_class'], axis=1)      #seperating dataset from target variable
X_test = test_set.drop(['wage_class'], axis=1)
Y_train = train_set['wage_class']                 #target variable
Y_test = test_set['wage_class']

In [19]:
#taking a look at the target variable
Y_train.value_counts()   


 <=50K    24720
 >50K      7841
Name: wage_class, dtype: int64

we can see from above that our dataset is imbalance. so for accuracy evaluation of the model we would 'roc_auc' method.<br>
also the value of the target variable is categorical in nature, so should convert it into numerical form befor applying it onto the model. 

In [20]:
Y_train = pd.Series([0 if x==' <=50K' else 1 for x in Y_train])
Y_test = pd.Series([0 if x==' <=50K.' else 1 for x in Y_test])

In [21]:
Y_train.head()

0    0
1    0
2    0
3    0
4    0
dtype: int64

In [22]:
Y_test.head()

0    0
1    0
2    1
3    1
4    0
dtype: int64

In [23]:
X_train.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  
0          2174             0              40   United-States  
1             0             0              13   United-States  
2             0             0              40   United-States  
3             0             0              40   United-States  
4             0             0              40            Cuba

In [24]:
X_train.iloc[0, 13]

' United-States'

### 4- Encoding datasets

many features in the above datasets are non-numeric in nature. so we must encode them into the numerical formate.

In [25]:
#finding unique values present in each categorical features

for col in X_train.columns:
    if X_train[col].dtypes == 'object':
        unique_cat = len(X_train[col].unique())
        print("{col} has {unique_cat} unique categories".format(col=col, unique_cat=unique_cat))

workclass has 9 unique categories
education has 16 unique categories
marital_status has 7 unique categories
occupation has 15 unique categories
relationship has 6 unique categories
race has 5 unique categories
sex has 2 unique categories
native_country has 42 unique categories


native_country has pretty large no. of unique categories, so lets see inside it for more clearity.

In [26]:
X_train['native_country'].value_counts().head(10)

 United-States    29170
 Mexico             643
 Philippines        198
 Germany            137
 Canada             121
 Puerto-Rico        114
 El-Salvador        106
 India              100
 Cuba                95
 England             90
Name: native_country, dtype: int64

here most of the observation belongs to only few categories, rest of them contains only few observation. so it would be a good idea to not create dummy for those categories instead we can put them into a seperate category named 'other'.

In [27]:
relevant_country = [' United-States', ' Mexico', ' Philippines', ' Germany', ' Canada', ' Puerto-Rico', ' El-Salvador', ' India',
                   ' Cuba', ' England']

In [28]:
X_train['native_country'] = [x if x in relevant_country else 'Other' for x in X_train['native_country']]
X_test['native_country'] = [x if x in relevant_country else 'Other' for x in X_test['native_country']]

print(X_train['native_country'].value_counts().sort_values(ascending=False))

 United-States    29170
Other              1787
 Mexico             643
 Philippines        198
 Germany            137
 Canada             121
 Puerto-Rico        114
 El-Salvador        106
 India              100
 Cuba                95
 England             90
Name: native_country, dtype: int64


In [29]:
X_train.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country       0
dtype: int64

In [30]:
X_train_dummies = pd.get_dummies(X_train, columns=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race',
                                                   'sex', 'native_country'])
X_train_dummies.head()

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  \
0                       0                     0                        0   
1                       0                     0                        0   
2                       0                     0                        0   
3                       0                     0                        0   
4                       0                     0                        0   

   workclass_ Private          ...           native_country_ Cuba  \
0                   0          ...                              0   
1                   0          ...                              0   
2                   1          ...                              0   
3                   1          ...                              0   
4                   1          ...                              1   

   native_country_ El-Salvador  native_country_ England  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   native_country_ Germany  native_country_ India  native_country_ Mexico  \
0                        0                      0                       0   
1                        0                      0                       0   
2                        0                      0                       0   
3                        0                      0                       0   
4                        0                      0                       0   

   native_country_ Philippines  native_country_ Puerto-Rico  \
0                            0                            0   
1                            0                            0   
2                            0                            0   
3                            0                            0   
4                            0                            0   

   native_country_ United-States  native_country_Other  
0                              1                     0  
1                              1                     0  
2                              1                     0  
3                              1                     0  
4                              0                     0  

[5 rows x 75 columns]

In [31]:
X_train_dummies.columns

Index(['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week', 'workclass_ Federal-gov', 'workclass_ Local-gov',
       'workclass_ Never-worked', 'workclass_ Private',
       'workclass_ Self-emp-inc', 'workclass_ Self-emp-not-inc',
       'workclass_ State-gov', 'workclass_ Without-pay', 'education_ 10th',
       'education_ 11th', 'education_ 12th', 'education_ 1st-4th',
       'education_ 5th-6th', 'education_ 7th-8th', 'education_ 9th',
       'education_ Assoc-acdm', 'education_ Assoc-voc', 'education_ Bachelors',
       'education_ Doctorate', 'education_ HS-grad', 'education_ Masters',
       'education_ Preschool', 'education_ Prof-school',
       'education_ Some-college', 'marital_status_ Divorced',
       'marital_status_ Married-AF-spouse',
       'marital_status_ Married-civ-spouse',
       'marital_status_ Married-spouse-absent',
       'marital_status_ Never-married', 'marital_status_ Separated',
       'marital_status_ Widowed', 'occu

In [32]:
X_train_dummies.isnull().any().sum()

0

In [33]:
X_test_dummies = pd.get_dummies(X_test, columns=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race',
                                                   'sex', 'native_country'])
X_test_dummies.head()
#X_test_dummies.columns

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week  \
0   25  226802              7             0             0              40   
1   38   89814              9             0             0              50   
2   28  336951             12             0             0              40   
3   44  160323             10          7688             0              40   
4   18  103497             10             0             0              30   

   workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  \
0                       0                     0                        0   
1                       0                     0                        0   
2                       0                     1                        0   
3                       0                     0                        0   
4                       0                     0                        0   

   workclass_ Private          ...           native_country_ Cuba  \
0                   1          ...                              0   
1                   1          ...                              0   
2                   0          ...                              0   
3                   1          ...                              0   
4                   0          ...                              0   

   native_country_ El-Salvador  native_country_ England  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   native_country_ Germany  native_country_ India  native_country_ Mexico  \
0                        0                      0                       0   
1                        0                      0                       0   
2                        0                      0                       0   
3                        0                      0                       0   
4                        0                      0                       0   

   native_country_ Philippines  native_country_ Puerto-Rico  \
0                            0                            0   
1                            0                            0   
2                            0                            0   
3                            0                            0   
4                            0                            0   

   native_country_ United-States  native_country_Other  
0                              1                     0  
1                              1                     0  
2                              1                     0  
3                              1                     0  
4                              1                     0  

[5 rows x 75 columns]

In [34]:
X_test_dummies.isnull().values.sum()

0

In [35]:
X_train_dummies.shape

(32561, 75)

In [36]:
X_test_dummies.shape

(16281, 75)

## Feature selection

In [37]:
import xgboost as xgb
from xgboost.sklearn import XGBClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler

from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.pipeline import make_pipeline
import pickle

In [38]:
model = XGBClassifier()
model.fit(X_train_dummies, Y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)

In [39]:
imp_feat = model.feature_importances_
imp_feat

array([0.15574963, 0.06259097, 0.1091703 , 0.14992721, 0.12372635,
       0.08879185, 0.00727802, 0.00436681, 0.        , 0.00291121,
       0.00727802, 0.02183406, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.00291121, 0.00291121, 0.        ,
       0.00291121, 0.        , 0.        , 0.00291121, 0.0014556 ,
       0.        , 0.00582242, 0.06113537, 0.        , 0.00727802,
       0.        , 0.0014556 , 0.0014556 , 0.        , 0.        ,
       0.02037846, 0.01455604, 0.00873362, 0.00436681, 0.01455604,
       0.        , 0.01455604, 0.00436681, 0.00436681, 0.00727802,
       0.0014556 , 0.        , 0.00291121, 0.0014556 , 0.01164483,
       0.00582242, 0.02765648, 0.        , 0.        , 0.0014556 ,
       0.        , 0.00727802, 0.01746725, 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.00291121, 0.0014556 , 0.        , 0.0014556 , 0.     

In [40]:
impp = [i for i in imp_feat if i > 0.0]
len(impp)

42

In [41]:
selection = SelectFromModel(model, threshold=0.001, prefit=True)

X_train_filt = selection.transform(X_train_dummies)
X_test_filt = selection.transform(X_test_dummies)

In [42]:
X_train_filt.shape

(32561, 42)

## Scaling :

In [43]:
sc = StandardScaler()
X_train_filtt = sc.fit_transform(X_train_filt)
X_test_filtt = sc.transform(X_test_filt)

C:\Users\sohini aich\Anaconda3\lib\site-packages\sklearn\utils\validation.py:475: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


## Model training :

### 1- Logistic Regression

In [44]:
pipe = make_pipeline(LogisticRegression(random_state=0))        #estimator model(logistic regression)

cv = StratifiedKFold(n_splits=20, random_state=42)      #defining the type of cross_validation

param_grid = {'logisticregression__C': [0.01, 0.1, 1, 10]}      #tunning parameters
    

grid = GridSearchCV(pipe, param_grid= param_grid, cv=cv, scoring="roc_auc") 

In [45]:
grid.fit(X_train_filtt, Y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=20, random_state=42, shuffle=False),
       error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('logisticregression', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'logisticregression__C': [0.01, 0.1, 1, 10]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [46]:
grid.best_params_

{'logisticregression__C': 1}

In [47]:
log_reg = pickle.dumps(grid)

## 2- Decision Tree :

In [48]:
pipe = make_pipeline(DecisionTreeClassifier(random_state= 45))          #estimator

cv = StratifiedKFold(n_splits=10, random_state=42)        #cross validation

param_grid = {'decisiontreeclassifier__criterion': ['gini', 'entropy'],
              'decisiontreeclassifier__max_depth': [3, 5, 7, 9, 13]}

grid = GridSearchCV(pipe, param_grid, cv=cv, scoring="roc_auc")

In [49]:
grid.fit(X_train_filtt, Y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=10, random_state=42, shuffle=False),
       error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('decisiontreeclassifier', DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=45,
            splitter='best'))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'decisiontreeclassifier__criterion': ['gini', 'entropy'], 'decisiontreeclassifier__max_depth': [3, 5, 7, 9, 13]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [50]:
grid.best_params_

{'decisiontreeclassifier__criterion': 'entropy',
 'decisiontreeclassifier__max_depth': 9}

In [51]:
Dectree_cla = pickle.dumps(grid)

## 3- Random Forest :

In [52]:
pipe = make_pipeline(RandomForestClassifier(n_estimators=500, random_state=123, n_jobs=-1))

cv = StratifiedKFold(n_splits=10, random_state=0)

param_grid = {'randomforestclassifier__max_features':['sqrt', 'log2', 10],
              'randomforestclassifier__max_depth':[9, 11, 13]}                 

grid = GridSearchCV(pipe, param_grid=param_grid, cv=cv, scoring="roc_auc")

In [53]:
grid.fit(X_train_filtt, Y_train)          #training 

GridSearchCV(cv=StratifiedKFold(n_splits=10, random_state=0, shuffle=False),
       error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('randomforestclassifier', RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=500, n_jobs=-1,
            oob_score=False, random_state=123, verbose=0, warm_start=False))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'randomforestclassifier__max_features': ['sqrt', 'log2', 10], 'randomforestclassifier__max_depth': [9, 11, 13]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [54]:
grid.best_params_

{'randomforestclassifier__max_depth': 13,
 'randomforestclassifier__max_features': 10}

In [55]:
Randfor_cla = pickle.dumps(grid)

## 4- Xgboost classifier :

In [56]:
pipe = make_pipeline(XGBClassifier(n_estimators=500, random_state=42, n_jobs=-1))

cv = StratifiedKFold(n_splits=10, random_state=0,)

param_grid = {'xgbclassifier__max_depth': [3, 5, 7],
              'xgbclassifier__learning_rate': [0.01, 0.1, 0.2]}

grid = GridSearchCV(pipe, param_grid=param_grid, cv=cv, scoring='roc_auc')

In [58]:
grid.fit(X_train_filtt, Y_train)

GridSearchCV(cv=StratifiedKFold(n_splits=10, random_state=0, shuffle=False),
       error_score='raise',
       estimator=Pipeline(memory=None,
     steps=[('xgbclassifier', XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=500,
       n_jobs=-1, nthread=None, objective='binary:logistic',
       random_state=42, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
       seed=None, silent=True, subsample=1))]),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'xgbclassifier__max_depth': [3, 5, 7], 'xgbclassifier__learning_rate': [0.01, 0.1, 0.2]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [65]:
grid.best_params_

{'xgbclassifier__learning_rate': 0.1, 'xgbclassifier__max_depth': 3}

In [59]:
xgbcla = pickle.dumps(grid)

## <u>Comparision between different models</u> :

In [60]:
log_reg = pickle.loads(log_reg)
Dectree_cla = pickle.loads(Dectree_cla)
Randfor_cla = pickle.loads(Randfor_cla)
xgbcla = pickle.loads(xgbcla)

In [61]:
print("""Logistic Regressor accuracy is {logg}
DecisionTree Classifier accuracy is {Dec}
RandomForest Classifier accuracy is {ran}
XGBoost Classifier accuracy is {xgb}""".format(logg=log_reg.score(X_test_filtt, Y_test),
                                               Dec=Dectree_cla.score(X_test_filtt, Y_test),
                                               ran=Randfor_cla.score(X_test_filtt, Y_test),
                                               xgb=xgbcla.score(X_test_filtt, Y_test)))

Logistic Regressor accuracy is 0.9052978452069326
DecisionTree Classifier accuracy is 0.9007075482054263
RandomForest Classifier accuracy is 0.9173319461930065
XGBoost Classifier accuracy is 0.9286006631258414


## Problem -1

We can predict whether a person makes over 50K a year with the accuracy approx 92%

## Problem -2

By feature selection we have performed above we know that some important factors are <br>
'age', 'fnlwgt', 'education_num', 'capital_gain', 'hours_per_week'

## Problem -3

by accuracy comparision performed above we can see that Xgboost classifier gives better result than other algorithm.